File: src/kiss/agents/vscode/src/brand.ts
// Author: Koushik Sen (ksen@berkeley.edu)
// Contributors:
// Koushik Sen (ksen@berkeley.edu)
// add your name here

/**
 * Product branding for the extension host.
 *
 * Mirrors `kiss.core.brand` on the Python side: the strings come from
 * `media/brand.json`, the one file a custom distribution replaces to
 * re-brand the product.  Read synchronously at module load because the
 * consumers are module-level constants (panel titles, notification
 * prefixes); the stock names are the fallback when the file is missing
 * or malformed, so the extension never fails to activate over branding.
 */

import * as fs from 'fs';
import * as path from 'path';

export interface Brand {
  productName: string;
  shortName: string;
  tagline: string;
}

const DEFAULT_BRAND: Brand = {
  productName: 'KISS Sorcar',
  shortName: 'KISS',
  tagline: 'Your AI assistant. Ask me anything!',
};

/** `media/brand.json`, resolved from `out/brand.js` in both dev and VSIX layouts. */
export const BRAND_FILE = path.join(__dirname, '..', 'media', 'brand.json');

/**
 * Read the brand strings from `file`, falling back key-by-key to the
 * stock KISS Sorcar names for anything missing, empty or not a string.
 */
export function loadBrand(file: string = BRAND_FILE): Brand {
  let raw: unknown;
  try {
    raw = JSON.parse(fs.readFileSync(file, 'utf-8'));
  } catch {
    return {...DEFAULT_BRAND};
  }
  const obj = (raw && typeof raw === 'object' ? raw : {}) as Record<
    string,
    unknown
  >;
  return {
    productName: pickString(obj, 'product_name', DEFAULT_BRAND.productName),
    shortName: pickString(obj, 'short_name', DEFAULT_BRAND.shortName),
    tagline: pickString(obj, 'tagline', DEFAULT_BRAND.tagline),
  };
}

function pickString(
  obj: Record<string, unknown>,
  key: string,
  fallback: string,
): string {
  const v = obj[key];
  return typeof v === 'string' && v ? v : fallback;
}

export const BRAND: Brand = loadBrand();
export const PRODUCT_NAME = BRAND.productName;
export const SHORT_NAME = BRAND.shortName;

/**
 * Fill the `{{PRODUCT_NAME}}`, `{{SHORT_NAME}}` and `{{TAGLINE}}` placeholders
 * in `text` (the twin of `kiss.core.brand.render_brand`, used on TIPS.md).
 * Other `{{...}}` tokens are left untouched.
 */
export function renderBrand(text: string, brand: Brand = BRAND): string {
  return text.replace(
    /\{\{(PRODUCT_NAME|SHORT_NAME|TAGLINE)\}\}/g,
    (_m, key: string) =>
      key === 'PRODUCT_NAME'
        ? brand.productName
        : key === 'SHORT_NAME'
          ? brand.shortName
          : brand.tagline,
  );
}

/**
 * Extension-relative path of the icon `media/<name>` as the installed
 * extension ships it.
 *
 * The VSIX build copies the icons to content-hashed names and records the
 * mapping in `media/hashed/index.json` (`scripts/hash-icons.js`) so that a
 * changed icon gets a fresh URL instead of the client's year-long cached
 * copy.  Icons shown at runtime (the editor-tab icon of a chat panel) go
 * through the same mapping; without the index (running from source) the
 * plain path is returned.
 */
export function mediaIconPath(extensionRoot: string, name: string): string {
  const plain = `media/${name}`;
  try {
    const index = JSON.parse(
      fs.readFileSync(
        path.join(extensionRoot, 'media', 'hashed', 'index.json'),
        'utf-8',
      ),
    );
    const hashed = index[plain];
    return typeof hashed === 'string' && hashed ? hashed : plain;
  } catch {
    return plain;
  }
}

File: src/kiss/core/brand.py
# Author: Koushik Sen (ksen@berkeley.edu)
# Contributors:
# Koushik Sen (ksen@berkeley.edu)
# add your name here
"""Product branding: the one place the product name and identity live.

Every user-visible name ("KISS Sorcar", the "KISS:" command prefix, the
agent's identity sentence in ``SYSTEM.md``) comes from
``src/kiss/agents/vscode/media/brand.json``.  The file sits in the
extension's ``media/`` directory because that is the single location
reachable by all three consumers: this Python package, the extension
host (``src/brand.ts``) and the shared ``chat.html`` template.  A custom
distribution re-brands the product by replacing that file (and the icon
files next to it) instead of editing source code: it drops its copies
into the git-ignored ``.brand/`` directory at the checkout root and
``install.sh`` swaps them in for the extension build (see "Brand
overlay" there), so the checked-in files always carry the stock brand.
"""

from __future__ import annotations

import json
import re
from pathlib import Path

_MEDIA_DIR = Path(__file__).resolve().parents[1] / "agents" / "vscode" / "media"
BRAND_FILE = _MEDIA_DIR / "brand.json"

DEFAULT_BRAND: dict[str, str] = {
    "product_name": "KISS Sorcar",
    "short_name": "KISS",
    "tagline": "Your AI assistant. Ask me anything!",
    "identity": (
        "You are KISS Sorcar, an AI Assistant and a general-purpose multi-model, "
        "multi-modal, multi-agent AI Agent Framework researched and developed by "
        "Koushik Sen (ksen@berkeley.edu)."
    ),
    "extension_description": (
        "The open-source AI coding agent that beats Cursor and Claude Code on "
        "Terminal Bench. Free, local, bring your own API key."
    ),
}

_PLACEHOLDER_RE = re.compile(r"\{\{(PRODUCT_NAME|SHORT_NAME|TAGLINE|IDENTITY)\}\}")


def load_brand(path: Path = BRAND_FILE) -> dict[str, str]:
    """Return the brand strings from *path*, falling back to the defaults.

    Missing or malformed files and missing keys fall back key-by-key to
    :data:`DEFAULT_BRAND`, so a partial ``brand.json`` (only a new
    ``product_name``) is enough to re-brand the product.  Non-string
    values are ignored the same way.
    """
    brand = dict(DEFAULT_BRAND)
    try:
        loaded = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, ValueError):
        return brand
    if isinstance(loaded, dict):
        for key in DEFAULT_BRAND:
            value = loaded.get(key)
            if isinstance(value, str) and value:
                brand[key] = value
    return brand


BRAND = load_brand()
PRODUCT_NAME = BRAND["product_name"]
SHORT_NAME = BRAND["short_name"]


def render_brand(text: str, brand: dict[str, str] = BRAND) -> str:
    """Fill the brand placeholders in *text*.

    Recognised tokens: ``{{PRODUCT_NAME}}``, ``{{SHORT_NAME}}``,
    ``{{TAGLINE}}`` and ``{{IDENTITY}}``.  Used on the prompt files
    (``SYSTEM.md``, ``SYSTEM_LITE.md``) whose
    identity sentence is brand-specific.  Unknown ``{{...}}`` tokens are
    left untouched so other templating in the same file is unaffected.
    """
    return _PLACEHOLDER_RE.sub(lambda m: brand[m.group(1).lower()], text)
